src/Core/KeyManagement/UserKey/Implementations/RotateUserAccountkeysCommand.cs

```diff
@@ -238,20 +188,70 @@ void UpdateUnlockMethods(RotateUserAccountKeysData model, User user, List<Update
         }
     }
 
+    private bool IsV2EncryptionUserAsync(User user)
+    {
+        // Returns whether the user is a V2 user based on the private key's encryption type.
+        ArgumentNullException.ThrowIfNull(user);
+        var isPrivateKeyEncryptionV2 = GetEncryptionType(user.PrivateKey) == EncryptionType.XChaCha20Poly1305_B64;
+        return isPrivateKeyEncryptionV2;
+    }
+
+    private async Task ValidateVerifyingKeyUnchanged(RotateUserAccountKeysData model, User user)
+    {
+        var currentSignatureKeyPair = await _userSignatureKeyPairRepository.GetByUserIdAsync(user.Id) ?? throw new InvalidOperationException("User does not have a signature key pair.");
+        if (model.AccountKeys.SignatureKeyPairData.VerifyingKey != currentSignatureKeyPair!.VerifyingKey)
+        {
+            throw new InvalidOperationException("The provided verifying key does not match the user's current verifying key.");
+        }
+    }
+
+    private static void ValidatePublicKeyEncryptionKeyPairUnchanged(RotateUserAccountKeysData model, User user)
+    {
+        var publicKey = model.AccountPublicKey ?? model.AccountKeys?.PublicKeyEncryptionKeyPairData?.PublicKey;
```

Comment:
The `?` does not need to be chained, when `PublicKeyEncryptionKeyPairData` cannot be null
```suggestion
        var publicKey = model.AccountPublicKey ?? model.AccountKeys?.PublicKeyEncryptionKeyPairData.PublicKey;
```